src/ray/object_manager/common.cc

```diff
@@ -0,0 +1,160 @@
+                 << "version: " << header->version << "\n"
+                 << "num_readers: " << header->num_readers << "\n"
+                 << "num_read_acquires_remaining: " << header->num_read_acquires_remaining
+                 << "\n"
+                 << "num_read_releases_remaining: " << header->num_read_releases_remaining
+                 << "\n"
+                 << "data_size: " << header->data_size << "\n"
+                 << "metadata_size: " << header->metadata_size << "\n";
+}
+
+void PlasmaObjectHeader::Init() {
+  // wr_mut is shared between writer and readers.
+  pthread_mutexattr_t mutex_attr;
+  pthread_mutexattr_init(&mutex_attr);
+  pthread_mutexattr_setpshared(&mutex_attr, PTHREAD_PROCESS_SHARED);
+  pthread_mutexattr_settype(&mutex_attr, PTHREAD_MUTEX_ERRORCHECK);
+  pthread_mutex_init(&wr_mut, &mutex_attr);
+
+  sem_init(&rw_semaphore, PTHREAD_PROCESS_SHARED, 1);
+
+  // Condition is shared between writer and readers.
+  pthread_condattr_t cond_attr;
+  pthread_condattr_init(&cond_attr);
+  pthread_condattr_setpshared(&cond_attr, PTHREAD_PROCESS_SHARED);
+  pthread_cond_init(&cond, &cond_attr);
+}
+
+void PlasmaObjectHeader::Destroy() {
+  RAY_CHECK(pthread_mutex_destroy(&wr_mut) == 0);
+  RAY_CHECK(pthread_cond_destroy(&cond) == 0);
+  RAY_CHECK(sem_destroy(&rw_semaphore) == 0);
+}
+
+// Get the data size of the plasma object.
+// This has to be called only when reader lock is acquired
+// via ReadAcquire.
+uint64_t PlasmaObjectHeader::GetDataSize() const {
+  RAY_CHECK_NE(num_read_releases_remaining, 0)
+      << "ReadAcquire has to be called before calling this method.";
+  return data_size;
+}
+
+void PlasmaObjectHeader::WriteAcquire(int64_t write_version,
+                                      uint64_t write_data_size,
+                                      uint64_t write_metadata_size,
+                                      int64_t write_num_readers) {
+  RAY_LOG(DEBUG) << "WriteAcquire. version: " << write_version << ", data size "
+                 << write_data_size << ", metadata size " << write_metadata_size
+                 << ", num readers: " << write_num_readers;
+  sem_wait(&rw_semaphore);
+  RAY_CHECK(pthread_mutex_lock(&wr_mut) == 0);
+  PrintPlasmaObjectHeader(this);
+
+  RAY_CHECK(num_read_acquires_remaining == 0);
+  RAY_CHECK(num_read_releases_remaining == 0);
+  RAY_CHECK(write_version == version + 1)
+      << "Write version " << write_version
+      << " is more than 1 greater than current version " << version
+      << ". Are you sure this is the only writer?";
+
+  version = write_version;
+  data_size = write_data_size;
+  metadata_size = write_metadata_size;
+  num_readers = write_num_readers;
+
+  RAY_LOG(DEBUG) << "WriteAcquire done";
+  PrintPlasmaObjectHeader(this);
+  RAY_CHECK(pthread_mutex_unlock(&wr_mut) == 0);
+}
+
+void PlasmaObjectHeader::WriteRelease(int64_t write_version) {
+  RAY_LOG(DEBUG) << "WriteRelease Waiting. version: " << write_version;
```

Comment:
Btw, this can be really spammy later... 

Wonder if we need a separate flag for these logs instead of DEBUG. Don't need to be done in this PR tho

Reply:
I think it's fine for now.